src/services/stringCompletions.ts

```diff
@@ -505,37 +522,44 @@ namespace ts.Completions.StringCompletions {
             }
         }
 
-        // check for a version redirect
-        const packageJsonPath = findPackageJson(baseDirectory, host);
-        if (packageJsonPath) {
-            const packageJson = readJson(packageJsonPath, host as { readFile: (filename: string) => string | undefined });
-            const typesVersions = (packageJson as any).typesVersions;
-            if (typeof typesVersions === "object") {
-                const versionResult = getPackageJsonTypesVersionsPaths(typesVersions);
-                const versionPaths = versionResult && versionResult.paths;
-                const rest = absolutePath.slice(ensureTrailingDirectorySeparator(baseDirectory).length);
-                if (versionPaths) {
-                    addCompletionEntriesFromPaths(result, rest, baseDirectory, extensions, versionPaths, host);
-                }
-            }
-        }
-
         return result;
     }
 
     function addCompletionEntriesFromPaths(result: NameAndKind[], fragment: string, baseDirectory: string, fileExtensions: readonly string[], paths: MapLike<string[]>, host: LanguageServiceHost) {
+        let pathResults: { results: NameAndKind[], matchedPattern: boolean }[] = [];
+        let matchedPathPrefixLength: number | undefined;
         for (const path in paths) {
             if (!hasProperty(paths, path)) continue;
             const patterns = paths[path];
             if (patterns) {
-                for (const { name, kind, extension } of getCompletionsForPathMapping(path, patterns, fragment, baseDirectory, fileExtensions, host)) {
-                    // Path mappings may provide a duplicate way to get to something we've already added, so don't add again.
-                    if (!result.some(entry => entry.name === name)) {
-                        result.push(nameAndKind(name, kind, extension));
+                const pathPattern = tryParsePattern(path);
+                if (!pathPattern) continue;
+                const isMatch = typeof pathPattern === "object" && isPatternMatch(pathPattern, fragment);
+                const isLongestMatch = isMatch && (matchedPathPrefixLength === undefined || pathPattern.prefix.length > matchedPathPrefixLength);
+                if (isLongestMatch) {
+                    // If this is a higher priority match than anything we've seen so far, previous results from matches are invalid, e.g.

```

Comment:
Can you give a code example? I assume you're trying to explain completions for

```ts
import * as bar from "bar$";
```

in the presence of the below `paths` object